### PR TITLE
Fix issue with anonymous enums

### PR DESCRIPTION
**Pull request by IDKWNTCMF**

### CODE REVIEW

File: server/src/clang-utils/SourceToHeaderMatchCallback.cpp [hunk 1]
```diff
@@ -209,6 +210,17 @@ void SourceToHeaderMatchCallback::generateInternal(const FunctionDecl *decl) con
         wrapperDecl = getOldStyleDeclarationAsString(decl, wrapperName);
     }
 
+    auto returnType = decl->getReturnType();
+    if (returnType->hasUnnamedOrLocalType()) {
+        if (auto EN = llvm::dyn_cast<clang::EnumDecl>(returnType->getAsTagDecl())) {
```
Comment (kichunya) [inline server/src/clang-utils/SourceToHeaderMatchCallback.cpp:215]: Change name EN on more understandable (for example enumStmt)
Also you can declare and define this variable before first if statement and use it in condition and in if body.

Comment (IDKWNTCMF) [reply server/src/clang-utils/SourceToHeaderMatchCallback.cpp]: Done

File: server/src/clang-utils/SourceToHeaderMatchCallback.cpp [hunk 2]
```diff
@@ -209,6 +210,16 @@ void SourceToHeaderMatchCallback::generateInternal(const FunctionDecl *decl) con
         wrapperDecl = getOldStyleDeclarationAsString(decl, wrapperName);
     }
 
+    auto returnType = decl->getReturnType();
+    auto enumDecl = llvm::dyn_cast_or_null<clang::EnumDecl>(returnType->getAsTagDecl());
+    if (returnType->hasUnnamedOrLocalType() && enumDecl) {
+        std::string returnTypeName = name + "_return_type";
```
Comment (ladisgin) [inline server/src/clang-utils/SourceToHeaderMatchCallback.cpp:216]: .

File: server/src/fetchers/FunctionDeclsMatchCallback.cpp [hunk 3]
```diff
@@ -42,6 +42,7 @@ void FunctionDeclsMatchCallback::run(const MatchFinder::MatchResult &Result) {
         }
         clang::QualType realReturnType = ClangUtils::getReturnType(FS, Result);
         methodDescription.returnType = ParamsHandler::getType(realReturnType, realReturnType, sourceManager);
+        methodDescription.returnType.replaceTypeNameIfUnnamed("enum " + methodName + "_return_type");
```
Comment (ladisgin) [inline server/src/fetchers/FunctionDeclsMatchCallback.cpp:45]: .

File: server/src/fetchers/FunctionDeclsMatchCallback.cpp [hunk 4]
```diff
@@ -72,6 +73,7 @@ void FunctionDeclsMatchCallback::run(const MatchFinder::MatchResult &Result) {
             methodDescription.accessSpecifier = getAcessSpecifier(FS);
         }
         methodDescription.returnType = ParamsHandler::getType(realReturnType, realReturnType, sourceManager);
+        methodDescription.returnType.replaceTypeNameIfUnnamed("enum " + methodName + "_return_type");
```
Comment (ladisgin) [inline server/src/fetchers/FunctionDeclsMatchCallback.cpp:76]: .

File: server/src/visitors/KleeAssumeReturnValueVisitor.cpp [hunk 5]
```diff
@@ -17,7 +17,10 @@ namespace visitor {
         functionCall = printer->constrFunctionCall(methodDescription, 0, "", false);
         additionalPointersCount = methodDescription.returnType.countReturnPointers();
         auto returnType = methodDescription.returnType.baseTypeObj();
-        printer->strDeclareVar(getActualTmpVarType(returnType).baseType(),
+        std::string type = (returnType.isUnnamed() && typesHandler->isEnum(returnType))
```
Comment (ladisgin) [inline server/src/visitors/KleeAssumeReturnValueVisitor.cpp:20]: .

File: server/src/clang-utils/SourceToHeaderMatchCallback.cpp [hunk 6]
```diff
@@ -264,6 +276,10 @@ void SourceToHeaderMatchCallback::generateWrapper(const FunctionDecl *decl) cons
         wrapperDecl = getOldStyleDeclarationAsString(decl, wrapperName);
     }
 
+    auto returnType = decl->getReturnType();
+    if (returnType->hasUnnamedOrLocalType() && returnType->isEnumeralType()) {
+        StringUtils::replaceFirst(wrapperDecl, "enum (anonymous)", "int");
```
Comment (ladisgin) [inline server/src/clang-utils/SourceToHeaderMatchCallback.cpp:281]: .

File: server/src/clang-utils/SourceToHeaderMatchCallback.cpp [hunk 7]
```diff
@@ -209,6 +210,16 @@ void SourceToHeaderMatchCallback::generateInternal(const FunctionDecl *decl) con
         wrapperDecl = getOldStyleDeclarationAsString(decl, wrapperName);
     }
 
+    auto returnType = decl->getReturnType();
+    auto enumDecl = llvm::dyn_cast_or_null<clang::EnumDecl>(returnType->getAsTagDecl());
+    if (returnType->hasUnnamedOrLocalType() && enumDecl) {
+        std::string returnTypeName = name + "_return_type";
+        renameDecl(enumDecl, returnTypeName);
+        print(enumDecl);
+        StringUtils::replaceFirst(wrapperDecl, "(anonymous)", returnTypeName);
```
Comment (ladisgin) [inline server/src/clang-utils/SourceToHeaderMatchCallback.cpp:219]: .

**Review by ladisgin — COMMENTED** [review]
Move duplicate code to separate functions

File: server/src/clang-utils/SourceToHeaderMatchCallback.cpp [hunk 8]
```diff
@@ -208,6 +209,13 @@ void SourceToHeaderMatchCallback::generateInternal(const FunctionDecl *decl) con
         curDecl = getOldStyleDeclarationAsString(decl, decoratedName);
         wrapperDecl = getOldStyleDeclarationAsString(decl, wrapperName);
     }
+    std::string enumReturnTypeName = PrinterUtils::getEnumReturnTypeName(name);
```
Comment (ladisgin) [inline server/src/clang-utils/SourceToHeaderMatchCallback.cpp:212]: Move into if block

File: server/src/utils/PrinterUtils.cpp [hunk 9]
```diff
@@ -97,6 +97,14 @@ namespace PrinterUtils {
         return methodName + "_return";
     }
 
+    std::string getDeclTypeName(const std::string& methodName) {
+        return methodName + "_return_type";
+    }
+
+    std::string getEnumReturnTypeName(const std::string& methodName) {
+        return "enum " + methodName + "_return_type";
```
Comment (ladisgin) [inline server/src/utils/PrinterUtils.cpp:105]: use `getDeclTypeName`

File: server/src/utils/PrinterUtils.cpp [hunk 10]
```diff
@@ -97,6 +97,14 @@ namespace PrinterUtils {
         return methodName + "_return";
     }
 
+    std::string getDeclTypeName(const std::string& methodName) {
```
Comment (ladisgin) [inline server/src/utils/PrinterUtils.cpp:100]: Bad naming, use smth like `getReturnMangledTypeName`

**Review by ladisgin — APPROVED** [review]

**Pull request merged**